### PR TITLE
Minimal MDNS - start a generic server implementation

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Minimal MDNS requires both client and server functionality (sending queries and receiving replies).

 #### Summary of Changes
Started generic server support:
  - bind to all interfaces (important for IPV6, where link-local addresses are per interface)
  - parsing queries and replies
  - added command line options for a server and client app
  - fixed some code review comments that my previous PR missed.

Still need to add reply generation methods. Code is more clean though

### CODE REVIEW

**Review by rwalker-apple — APPROVED** [review]

File: src/lib/support/BufBound.h [hunk 1]
```diff
@@ -68,6 +69,7 @@ class BufBound
         }
         return mWritten;
     }
+    size_t Put8(uint8_t c) { return Put(c); }
```
Comment (rwalker-apple) [inline src/lib/support/BufBound.h:72]: I wonder if BufBound should return itself like BufReader does.  returning bytes written is generally uninteresting to clients

Comment (andy31415) [reply src/lib/support/BufBound.h:72]: +1 to that. I did not want to change the interface more, however would absolutely support a PR changing things to return *this. May make other code cleaner as a result as well.

File: src/lib/mdns/minimal/Server.cpp [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                                     const chip::Inet::IPPacketInfo * info)
+{
+    chip::System::AutoFreePacketBuffer autoFree(buffer);
+
+    ServerBase * srv = static_cast<ServerBase *>(endPoint->AppState);
+    if (!srv->mDelegate)
+    {
+        return;
+    }
+    mdns::Minimal::BytesRange data(mdns::Minimal::BytesRange(buffer->Start(), buffer->Start() + buffer->DataLength()));
```
Comment (mspang) [inline src/lib/mdns/minimal/Server.cpp:223]: remove this redundant copy?

File: src/lib/mdns/minimal/Server.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#include <inet/IPAddress.h>
+#include <inet/InetInterface.h>
+#include <inet/UDPEndPoint.h>
+
+#include "BytesRange.h"
+
+namespace mdns {
+namespace Minimal {
+
+class ListenIterator
```
Comment (mspang) [inline src/lib/mdns/minimal/Server.h:30]: Provide some class comments here.

File: src/lib/mdns/minimal/Server.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    virtual ~ServerDelegate() {}
+
+    // Callback of when a query is received
+    virtual void OnQuery(const BytesRange & data, const chip::Inet::IPPacketInfo * info) = 0;
+
+    // Callback of when a response is received
+    virtual void OnResponse(const BytesRange & data, const chip::Inet::IPPacketInfo * info) = 0;
+};
+
+// Defines an mDNS server that listends on one or more interfaces.
```
Comment (mspang) [inline src/lib/mdns/minimal/Server.h:52]: typo: listends -> listens

**Review by mspang — APPROVED** [review]

File: src/lib/mdns/minimal/Server.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/mdns/minimal/Server.cpp:223]: Heh ... copy and paste got to me. Fixed.

**Comment by andy31415** [comment]
@saurabhst @BroderickCarlin @hawk248 @jelderton ?

**Review by jelderton — APPROVED** [review]

**Pull request merged**